README.md
# Flutter - Intro Screens

A Flutter based Reusable Intro Screens component that provides Application Startup Screens with configurable items which can be used in any Flutter application.


## Table of contents
- [Flutter Support](#Flutter-support)
- [Demo](#demo)
- [Features](#features)
- [Getting started](#getting-started)
- [Usage](#usage)
- [Methods](#methods)
- [Want to Contribute?](#want-to-contribute)
- [Need Help / Support?](#need-help) 
- [Collection of Components](#collection-of-Components)
- [Changelog](#changelog)
- [License](#license)
- [Keywords](#Keywords)


## Flutter Support

Version - Flutter 1.17 (stable)

We have tested our program in above version, however you can use it in other versions as well.


## Demo
[![](introScreen.gif)](https://github.com/weblineindia/Flutter-Intro-Screens/blob/master/introScreen.gif)

------

## Features

* Introduction Screens handled with navigation using stateless widget.
* Introduction Screens with less code of implementation and configurable UI options.


## Getting started

* Download this sample project and import app_startup_ui_widget.dart Widget file in your Flutter App. 
* Update images based on your requirements. 


## Usage

Setup process is described below to integrate in project.

## Configure Intro Screens View in your widget by following below method steps

### Methods

Step 1.Add  'provider' package to pubspec.yaml

Step 2. Add MultiProvider in Your application widget like below and provider StartupScreenNavigationProvider in providers list.
       
    // Put BottomNavigatorProvider as mention below
    class MyApp extends StatelessWidget {
      @override
      Widget build(BuildContext context) {
        final MaterialColor colorCustom = MaterialColor(0xFFFACA58, color);
        return MultiProvider(
          providers: [
            ChangeNotifierProvider.value(
              value: StartupScreenNavigationProvider(),
            )
          ],
          child: MaterialApp(
            title: 'Flutter Demo',
            home: ...,
          ),
        );
      }
    }
       

Step 3. User following StartUpScreen constructor for adding Intro Screen as user landing screen.             
       
         List<PageItem> arrPageItems = [];
         Color pageIndicatorColor;
         Color pageIndicatorSelectedColor;
         Color buttonsColor;
         Color buttonsTextColor;
         Function(BuildContext, int) nextPageCallback;
         Function(BuildContext) skipScreenCallback;
         

         // ignore: sort_constructors_first
         StartUpScreen(
             {this.arrPageItems,
             this.pageIndicatorColor,
             this.pageIndicatorSelectedColor,
             this.buttonsColor,
             this.buttonsTextColor,
             this.skipScreenCallback,
             this.nextPageCallback});
        

Step 4. Pass PageItems to StartUpScreen widget. 

      this.arrPageItems = [] // Pass PageItem in StartUpScreen widget for displaying and Use 'PageItem' helper class to create PageItems 

     //Use our helper class to create pageItems for StartUpScreen
     class PageItem {
       String imgPath;
       // ignore: sort_constructors_first
       PageItem({this.imgPath});
     }


Step 5.if you want to set various colors to StartUpScreen you can configure as below.

    this.pageIndicatorColor // To set page indicators unselected color
    this.pageIndicatorSelectedColor // To set page indicators selected color
    this.buttonsColor // To set buttons color in StartUpScreen Widget
    this.buttonsTextColor // To set buttons text color in StartUpScreen Widget
    
Step 6.if you want to Handle callbacks of StartUpScreen.
    
    // Use below parameter of StartUpScreen widget for handle skip button callback
    skipScreenCallback: (BuildContext context) {
      Navigator.pushReplacement(
          context, MaterialPageRoute(builder: (context) => YOUR_SCREEN()));
    } 
    
    // Use below parameter of StartUpScreen widget for handle next and last page callback of StartUpScreen 
    nextPageCallback: (BuildContext context, int page) {
      if (page == arrItems.length - 1) {
        Navigator.pushReplacement(context,
            MaterialPageRoute(builder: (context) => YOUR_SCREEN()));
       }
       else {...}
    }
    
Step 7. Adding StartUpScreen in your main app widget
    
    MaterialApp(
      title: 'Flutter Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: colorCustom,
      ),
      home: StartUpScreen(
        arrPageItems: arrItems,
        pageIndicatorColor: Colors.grey,
        pageIndicatorSelectedColor: colorCustom,
        buttonsColor: colorCustom,
        buttonsTextColor: Colors.white,
        skipScreenCallback: (BuildContext context) {
          Navigator.pushReplacement(
              context, MaterialPageRoute(builder: (context) => YOUR_SCREEN()));
        },
        nextPageCallback: (BuildContext context, int page) {
          if (page == arrItems.length - 1) {
            Navigator.pushReplacement(context,
                MaterialPageRoute(builder: (context) => YOUR_SCREEN()));
          }
        },
      ),
    )
  
    
------

## Want to Contribute?

- Created something awesome, made this code better, added some functionality, or whatever (this is the hardest part).
- [Fork it](http://help.github.com/forking/).
- Create new branch to contribute your changes.
- Commit all your changes to your branch.
- Submit a [pull request](http://help.github.com/pull-requests/).

------

## Need Help? 
If you need help then feel free to contact our [Flutter developers](https://www.weblineindia.com/flutter-cross-platform-mobile-app-development.html).

 ------
 
## Collection of Components
 We have built many other components and free resources for software development in various programming languages. Kindly click here to view our [Free Resources for Software Development.](https://www.weblineindia.com/software-development-resources.html)
 
------

## Changelog
Detailed changes for each release are documented in [CHANGELOG](./CHANGELOG).

## License

[MIT](LICENSE)

[mit]: https://github.com/weblineindia/Flutter-Intro-Screens/blob/master/LICENSE

## Keywords
Flutter-Intro-screens, Introduction-screens, App-startup-screen, App-onboarding-screen


CHANGELOG
1.0.0 (July 22, 2020)
Initial release - Sample source code for Flutter Reusable Introduction Screens Component.


lib/home_screen.dart
import 'package:flutter/material.dart';

// ignore: must_be_immutable
class HomeScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
          centerTitle: true,
          iconTheme: const IconThemeData(color: Colors.white),
          title: const Text('Home', style: TextStyle(color: Colors.white)),
          brightness: Brightness.dark),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.start,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            Image.asset(
              'assets/logo.png',
              height: MediaQuery.of(context).size.height * 0.20,
              width: 350.0,
            ),
            SizedBox(
              height: MediaQuery.of(context).size.height * 0.15,
              width: 50,
            ),
            const Text('Welcome',
                style: TextStyle(
                  fontSize: 33,
                  fontWeight: FontWeight.w900,
                )),
          ],
        ),
      ),
    );
  }
}


lib/app_startup_ui_widget.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

// ignore: must_be_immutable
class StartUpScreen extends StatelessWidget {
  List<PageItem> arrPageItems = [];
  Color pageIndicatorColor;
  Color pageIndicatorSelectedColor;
  Color buttonsColor;
  Color buttonsTextColor;
  Function(BuildContext, int) nextPageCallback;
  Function(BuildContext) skipScreenCallback;
  PageController controller = PageController(initialPage: 0);

  // ignore: sort_constructors_first
  StartUpScreen(
      {this.arrPageItems,
      this.pageIndicatorColor,
      this.pageIndicatorSelectedColor,
      this.buttonsColor,
      this.buttonsTextColor,
      this.skipScreenCallback,
      this.nextPageCallback});

  @override
  Widget build(BuildContext context) {
    return Container(
      child: Container(
        child: Stack(
          children: [
            _createPageBuilder(context),
            Column(
              mainAxisAlignment: MainAxisAlignment.end,
              crossAxisAlignment: CrossAxisAlignment.center,
              children: [
                _createPageIndicator(context),
                _createButtonsContainer(context),
              ],
            )
          ],
        ),
      ),
    );
  }

  Widget _createButtonsContainer(BuildContext context) {
    final height = MediaQuery.of(context).size.height;
    return Align(
      alignment: Alignment.bottomCenter,
      child: Container(
        padding: const EdgeInsets.all(10.0),
        margin: EdgeInsets.only(bottom: height * 0.10),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: [
            RaisedButton(
              child: Text(
                'Skip',
                style: TextStyle(color: buttonsTextColor),
              ),
              onPressed: () {
                skipScreenCallback(context);
              },
              color: buttonsColor,
              shape: RoundedRectangleBorder(
                borderRadius: BorderRadius.circular(20),
              ),
            ),
            Consumer<StartupScreenNavigationProvider>(
                builder: (ctx, item, child) {
              return RaisedButton(
                child: Text(
                  item.pageSelectedIndex == arrPageItems.length - 1
                      ? 'Finish'
                      : 'Continue',
                  style: TextStyle(color: buttonsTextColor),
                ),
                onPressed: () {
                  if (item.pageSelectedIndex != arrPageItems.length - 1) {
                    controller.animateToPage(
                      item.pageSelectedIndex + 1,
                      duration: const Duration(milliseconds: 300),
                      curve: Curves.linear,
                    );
                  } else {}
                  nextPageCallback(context, item.pageSelectedIndex);
                },
                color: buttonsColor,
                shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(20),
                ),
              );
            })
          ],
        ),
      ),
    );
  }

  PageView _createPageBuilder(BuildContext context) {
    final StartupScreenNavigationProvider provider =
        Provider.of<StartupScreenNavigationProvider>(context, listen: false);
    return PageView.builder(
      physics: const ClampingScrollPhysics(),
      controller: controller,
      itemBuilder: (ctx, index) => _pageBuilder(ctx, arrPageItems[index]),
      itemCount: arrPageItems.length,
      onPageChanged: (pageIndex) {
        provider.setPageSelectedIndex(pageIndex);
      },
    );
  }

  Widget _createPageIndicator(BuildContext context) {
    final height = MediaQuery.of(context).size.height;
    return Consumer<StartupScreenNavigationProvider>(
        builder: (ctx, item, child) {
      return Stack(
        alignment: AlignmentDirectional.bottomCenter,
        children: <Widget>[
          Align(
            alignment: Alignment.bottomCenter,
            child: Container(
              margin: EdgeInsets.only(bottom: height * 0.05),
              child: Row(
                mainAxisSize: MainAxisSize.min,
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  for (int i = 0; i < arrPageItems.length; i++)
                    if (i == item.pageSelectedIndex)
                      _circleIndicator(true)
                    else
                      _circleIndicator(false),
                ],
              ),
            ),
          ),
        ],
      );
    });
  }

  Widget _circleIndicator(bool isActive) {
    return AnimatedContainer(
      duration: const Duration(milliseconds: 150),
      margin: const EdgeInsets.symmetric(horizontal: 8),
      height: isActive ? 12 : 8,
      width: isActive ? 12 : 8,
      decoration: BoxDecoration(
          color: isActive ? pageIndicatorSelectedColor : pageIndicatorColor,
          borderRadius: const BorderRadius.all(Radius.circular(12))),
    );
  }

  Widget _pageBuilder(BuildContext context, PageItem page) {
    return Container(
      height: double.infinity,
      width: double.infinity,
      decoration: const BoxDecoration(color: Colors.cyan),
      child: Image.asset(
        page.imgPath,
        fit: BoxFit.cover,
      ),
    );
  }
}

class PageItem {
  String imgPath;
  // ignore: sort_constructors_first
  PageItem({this.imgPath});
}

class StartupScreenNavigationProvider extends ChangeNotifier {
  int pageSelectedIndex = 0;
  void setPageSelectedIndex(int sIndex) {
    pageSelectedIndex = sIndex;
    notifyListeners();
  }
}


lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_tab_bar/home_screen.dart';
import 'package:provider/provider.dart';
import 'app_startup_ui_widget.dart';

/*
Author Name  :  WeblineIndia  |  https://www.weblineindia.com/
For more such software development components and code libraries, visit us at
https://www.weblineindia.com/software-development-resources.html 
Our Github URL : https://github.com/weblineindia
*/

void main() => runApp(MyApp());

// ignore: must_be_immutable
class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  Map<int, Color> color = {
    50: const Color.fromRGBO(250, 202, 88, .1),
    100: const Color.fromRGBO(250, 202, 88, .2),
    200: const Color.fromRGBO(250, 202, 88, .3),
    300: const Color.fromRGBO(250, 202, 88, .4),
    400: const Color.fromRGBO(250, 202, 88, .5),
    500: const Color.fromRGBO(250, 202, 88, .6),
    600: const Color.fromRGBO(250, 202, 88, .7),
    700: const Color.fromRGBO(250, 202, 88, .8),
    800: const Color.fromRGBO(250, 202, 88, .9),
    900: const Color.fromRGBO(250, 202, 88, 1),
  };

  List<PageItem> arrItems = [
    PageItem(imgPath: 'assets/image1.jpg'),
    PageItem(imgPath: 'assets/image2.jpg'),
    PageItem(imgPath: 'assets/image3.jpg'),
  ];

  @override
  Widget build(BuildContext context) {
    final MaterialColor colorCustom = MaterialColor(0xFFFACA58, color);
    return MultiProvider(
      providers: [
        ChangeNotifierProvider.value(value: StartupScreenNavigationProvider())
      ],
      child: MaterialApp(
        title: 'Flutter Demo',
        debugShowCheckedModeBanner: false,
        theme: ThemeData(
          primarySwatch: colorCustom,
        ),
        home: StartUpScreen(
          arrPageItems: arrItems,
          pageIndicatorColor: Colors.grey,
          pageIndicatorSelectedColor: colorCustom,
          buttonsColor: colorCustom,
          buttonsTextColor: Colors.white,
          skipScreenCallback: (BuildContext context) {
            Navigator.pushReplacement(
                context, MaterialPageRoute(builder: (context) => HomeScreen()));
          },
          nextPageCallback: (BuildContext context, int page) {
            if (page == arrItems.length - 1) {
              Navigator.pushReplacement(context,
                  MaterialPageRoute(builder: (context) => HomeScreen()));
            }
          },
        ),
      ),
    );
  }
}
